enhancements/installer/openstack-ipi-failure-domains.md

```diff
@@ -0,0 +1,340 @@
+
+The OpenShift administrators provide a list of `failureDomains` and their information:
+
+```yaml
+platform:
+  openstack:
+    ...
+    failureDomains:
+    - name: rack1
+      computeZone: az1
+      storageZone: cinder-az1
+      subnet: d7ffd3d8-87e1-481f-a818-ff2f17787d40
+    - name: rack2
+      computeZone: az2
+      storageZone: cinder-az2
+      subnet: f584ce8c-bb54-422b-8e11-6a530be4c23e
+    - ...
+```
+There might be additional options in the domains, like flavors or BGP peers, but for now we will cover the minimum.
+We reserve ourselves the possibility to expand the definition of failure domains in the future.
+
+Then they'll choose what Failure Domains to use when deploying the control plane and compute nodes by setting the `failureDomain` field in the machine pools:
+
+```yaml
+controlPlane:
+  name: master
+  platform:
+    openstack:
+      failureDomains:
+      - rack1
+      - rack2
+      - rack3
+  replicas: 3 // this will deploy 1 master in each domain
+compute:
+- name: worker
+  platform:
+    openstack:
+      ...
+      failureDomains:
+      - rack1
+      - rack2
+      - rack3
+      - rack4
+      - rack5
+      rootVolume:
+        size: 30
+        type: performance
+  replicas: 10 // this will deploy 2 workers per rack
+```
+
+The installer creates a machineSet per failure domain passed to the compute machinePool.
+
+The machineSet for the compute's `rack1` failure domain will look like the following:
+```yaml
+apiVersion: machine.openshift.io/v1beta1
+kind: MachineSet
+metadata:
+  labels:
+    machine.openshift.io/cluster-api-cluster: <infrastructure_ID>
+    machine.openshift.io/cluster-api-machine-role: worker
+    machine.openshift.io/cluster-api-machine-type: worker
+  name: <infrastructure_ID>-worker-rack1
+  namespace: openshift-machine-api
+spec:
+  replicas: 2
+  selector:
+    matchLabels:
+      machine.openshift.io/cluster-api-cluster: <infrastructure_ID>
+      machine.openshift.io/cluster-api-machineset: <infrastructure_ID>-worker-rack1
+  template:
+    ...
+    spec:
+      metadata:
+      providerSpec:
+        value:
+          apiVersion: openstackproviderconfig.openshift.io/v1alpha1
+          cloudName: openstack
+          flavor: <nova_flavor>
+          kind: OpenstackProviderSpec
+          networks:
+            - subnets:
+              - uuid: d7ffd3d8-87e1-481f-a818-ff2f17787d40
+          availabilityZone: az1
+          rootVolume:
+            diskSize: 30
+            volumeType: performance
+            availabilityZone: cinder-az1
+          ...
+```
+
+The installer will create the control plane machines in a way that they can easily be adopted by the [ControlPlane machineset](https://github.com/openshift/enhancements/blob/master/enhancements/machine-api/control-plane-machine-set.md).
+
+When deploying nodes on separate subnets, the user will need to update the `machineNetwork` option to list all the possible CIDRs on which the machines can be:
```

Comment:
let's focus on control plane:
e.g. a new failure domain is added to the infra (with a new subnet, etc)
* if we want a control plane machine to be moved into it, then we follow the standard [replacement](https://docs.openshift.com/container-platform/4.10/backup_and_restore/control_plane_backup_and_restore/replacing-unhealthy-etcd-member.html) procedure after changing the MachineConfig of one master that we want to move. This is not ideal but until we get MachineSets for Control Plane this is the only thing that I see would work.
* We have to manually update the security groups for the masters, not ideal but this would be documented. This should be handle by MAO/MAPO at some point? This is a well known issue, even today, where the security groups aren't managed by any operator right now.

Reply:
I'll add that to open questions btw.

Reply:
As far as I can tell, `machineNetwork` isn't used outside of the installer, and the sole purpose of `machineNetwork` is to generate the initial security group rules.

Now, I'm not aware of an official way to manage the security group rules after the installation, for instance in case of new security rules needed after an upgrade, or in this case updating the list of CIDR the machines can be on. I suppose the responsibility currently falls on the customer, but I imagine an operator could easily manage the security rules and keep them in sync based on a cluster property.

@patrickdillon do you know how the different platforms manage their security rules?